### ISSUE TITLE
Export and restoreable all settings

### ISSUE THREAD

**Issue by NSTAdventure** [bromite/bromite#1331]
yes

### Is your feature request related to privacy?
Yes:
Chrome//flag have some setting that can be tweaked by user the only issues is if data deleted or bromite reinstalled then you need re config those setting one by one very annoying

### Is there a patch available for this feature somewhere?
No? 🤔

### Describe the solution you would like
Able to export and restore all setting bromite included chrome//flag

### Describe alternatives you have considered
I found no alternative

**Comment by kokukyocho** [comment]
Hello @NSTAdventure!

I am a bot :robot:

Thanks for submitting this issue!
I noticed that it is missing the template, please edit the issue to match the template for either a [bug](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) or a [feature request](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/feature_request.md); the issue will be automatically re-opened afterwards.

If this is not an issue nor a feature request and you want to start a discussion then you can use [GitHub Discussions](https://github.com/bromite/bromite/discussions).

**Tip**: if you use GitHub in Desktop mode the template will be proposed to you when submitting issues.

**Comment by csagan5** [comment]
This would be quite a bit complex to implement; @uazo technically it would have to cover the current profile preferences only, right?

**Comment by uazo** [comment]
> technically it would have to cover the current profile preferences only, right?

the preferences seem to me to be saved in two files:

- /data/data/org.bromite.bromite/app_chrome/Local State
- /data/data/org.bromite.bromite/app_chrome/Default/Preferences

In the first you can find the preferences not linked to any profile (i.e. the flags, the metrics, the proxy, etc.), in the other the rest, per user (even if there is only one user in bromite). both are plain text in json format.
I have never tried, but I guess that to restore it is enough to replace those files, as they are not protected.
it should be done before starting the native, java side. so, basically, for the restore, java should check some file placed in a predefined location by the ui, and, at restart, before starting the native, replace those files with those made from the backup.
from a security point of view it worries me a little, knowing that my preferences can be tampered with from the outside.
I would prefer first that the settings are accessible are from a pincode or from the footprints and then do the backup / restore.
but this is the "dirty" method, I think.

the clean method i guess is to use chrome's built-in backup: [java](https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/ChromeBackupAgentImpl.java) [native](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/android/chrome_backup_agent.cc)
should be seen how it works and reused in bromite. it was something I discovered time ago and I wanted to see.
I guess that's best, considering that backups are also done for bookmarks, passwords and everything in between.

**Comment by csagan5** [comment]
> from a security point of view it worries me a little, knowing that my preferences can be tampered with from the outside.

Only the user can; other apps cannot tamper with these files.



> the clean method i guess is to use chrome's built-in backup: [java](https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/ChromeBackupAgentImpl.java) [native](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/android/chrome_backup_agent.cc)
> should be seen how it works and reused in bromite. it was something I discovered time ago and I wanted to see.
> I guess that's best, considering that backups are also done for bookmarks, passwords and everything in between.

This one does not backup flags and proxy settings (which is what this issue is about, including profile preferences) but the whole profile instead.

It is used here: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/chrome_browser_main_android.cc;drc=6fdd6d4796fc076fd06f31bce5ea2426e478620c;bpv=1;bpt=1;l=66

So the purpose is for the Android backup system.

Would it be doable to perform a backup of flags, proxy settings and profile preferences (and a restore) using an import/export functionality from the browser itself (and not using the Android backup functionality instead)? Perhaps there is a reason why it is done via Android backup e.g. you cannot change files while they are open.

**Comment by uazo** [comment]
> Only the user can; other apps cannot tamper with these files.

yes, but theoretically as usual. in any case we would open a "backdoor" that currently does not exist.

> Would it be doable to perform a backup of flags, proxy settings and profile preferences (and a restore) using an import/export functionality from the browser itself 

yes, LocalState is handled internally by the same preferences API.

> Perhaps there is a reason why it is done via Android backup e.g. you cannot change files while they are open.

no, I do not think. the preferences are all in memory and saved on disk occosionally or on specific api request (`CommitPendingWrite`)

> So the purpose is for the Android backup system.

I'm sorry not to be more precise, but looking at the tests it would seem that we can intercept the calls and produce a file rather than sending it to the android backup system.
https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/junit/src/org/chromium/chrome/browser/ChromeBackupAgentTest.java

**Comment by NSTAdventure** [comment]
> This would be quite a bit complex to implement; @uazo technically it would have to cover the current profile preferences only, right?

Anyway did you get what i mean?

Just in case if not:
I mean you can backup those things:
![Screenshot_2021-08-12-15-47-31-90_ed7821ed99ee442dec4ede205e7971d6](https://user-images.githubusercontent.com/50813528/129167539-e7d8e572-aa94-4c35-ae18-a9ff004962f4.jpg)

![Screenshot_2021-08-12-15-47-34-43_ed7821ed99ee442dec4ede205e7971d6](https://user-images.githubusercontent.com/50813528/129167565-ac266757-943b-46bf-9053-d6e526840105.jpg)

![Screenshot_2021-08-12-15-48-19-81_ed7821ed99ee442dec4ede205e7971d6](https://user-images.githubusercontent.com/50813528/129167643-923e6e13-9c77-4234-929b-6554d6a98b32.jpg)

Especially chrome://flags or experimental setting
And website

@csagan5
@uazo 

Best Regards
NST_Adventure

**Comment by csagan5** [comment]
> > Only the user can; other apps cannot tamper with these files.
> 
> yes, but theoretically as usual. in any case we would open a "backdoor" that currently does not exist.

It is not theoretical: it is practical. If you can prove that other apps can break the sandbox then you can apply for an Android security bounty for the sandbox breakage, otherwise please do not spread misinformation. We have already discussed this on #1273.

There is no backdoor, users can change those files already as they please.

> > So the purpose is for the Android backup system.
> 
> I'm sorry not to be more precise, but looking at the tests it would seem that we can intercept the calls and produce a file rather than sending it to the android backup system.

We do not need this logic, this feature request is about flags, proxy settings and profile preferences, not everything else.

**Comment by uazo** [comment]
> If you can prove that other apps can break the sandbox 
> otherwise please do not spread misinformation.

no, absolutely not what I want. 
what I think (and maybe I have explained it wrong) is that in a rooted system, access to those files can be allowed and therefore the restore phase should somehow be made more secure.

> not everything else

sure, but if it's free, it's better not?

**Comment by csagan5** [comment]
> > If you can prove that other apps can break the sandbox
> > otherwise please do not spread misinformation.
> 
> no, absolutely not what I want.
> what I think (and maybe I have explained it wrong) is that in a rooted system, access to those files can be allowed and therefore the restore phase should somehow be made more secure.

You have written: the fact that apps cannot tamper with other apps' data is theoretical.

I have written: this is false, because the Android sandbox exists.

You are now writing: "in a rooted system, access to those files can be allowed and therefore the restore phase should somehow be made more secure."

The access is allowed by the SuperSU app (or similar), on-demand, and still requires user interaction.

If you refer to some app circumventing SuperSU on an already-rooted phone: this is out of scope for this project, I will not merge anything related to that. It is impossible to add any security feature to protect users in such scenario, I have already done my research and I invite you to do that as well.

> > not everything else
> 
> sure, but if it's free, it's better not?

Not if it increases the work needed to deliver the subset discussed here.

**Comment by uazo** [comment]
> If you refer to some app circumventing SuperSU on an already-rooted phone: this is out of scope for this project

I think of who has the rooted phone but with little familiarity with what it means

> It is impossible to add any security feature to protect users in such scenario

in the hypothesis that a replacement of those files at startup can be developed, we could first ask for confirmation before replacing them, perhaps with a nice warning message. better if pin or fingerprint reading is required, but I just don't know how simple it is to do.

> Not if it increases the work needed to deliver the subset discussed here.

I checked the code, it doesn't do what I thought at all: it backup a few specific preferences, among other things only booleans. however it can be used as a basis, if after a check, turns out to be simpler than the method of replacing files, but I don't think.
Also I don't know how to retrieve the list of saved preference keys.

So, two hypotheses

A) 
Backup: "UI" with SAF for saving files in zip format
Restore: always with SAF choice of the zip and saving in a default directory + restart of bromite. Upon restart, request for confirmation and replacement of files.

B)
By modifying the chrome backup and without using AndroidBackup, preparation of a file with all preferences. the restore applies the preferences via api (or a wired subset of it if you can't get a list at runtime)

for now the a) seems simpler to me

**Comment by csagan5** [comment]
> > It is impossible to add any security feature to protect users in such scenario
> 
> in the hypothesis that a replacement of those files at startup can be developed, we could first ask for confirmation before replacing them, perhaps with a nice warning message. better if pin or fingerprint reading is required, but I just don't know how simple it is to do.

If the phone is rooted and another app roots it again for its own malicious purposes it will be able to circumvent anything we put in place and thus I will not merge any such functionality which would give a fake sense of security.



> A)
> Backup: "UI" with SAF for saving files in zip format
> Restore: always with SAF choice of the zip and saving in a default directory + restart of bromite. Upon restart, request for confirmation and replacement of files.
> 
> B)
> By modifying the chrome backup and without using AndroidBackup, preparation of a file with all preferences. the restore applies the preferences via api (or a wired subset of it if you can't get a list at runtime)

The android backup is something that can be used only when the app is not running: it is something the user requests, not the app.

**Comment by csagan5** [comment]
There is a work-in-progress by @uazo for this issue in #1500; should we continue with this or first look into #244 (or similar)?

**Comment by uazo** [comment]
the problem with that patch is how to encrypt the data .. otherwise it's fine

> should we continue with this or first look into https://github.com/bromite/bromite/issues/244 (or similar)?

every time I reread this issue I think about it ...
I don't know if #244 is worth doing that way, it forces the use of software outside the browser, perhaps even not managed by the user... I have to think about it a little longer

**Comment by csagan5** [comment]
> the problem with that patch is how to encrypt the data .. otherwise it's fine

I already replied to this. Either no encryption or ask user to provide one.

> > should we continue with this or first look into #244 (or similar)?
> 
> every time I reread this issue I think about it ... I don't know if #244 is worth doing that way, it forces the use of software outside the browser, perhaps even not managed by the user... I have to think about it a little longer

It is fundamentally different from this issue. #244 is mostly to use the same profile across multiple devices.

**Comment by uazo** [comment]
> Either no encryption or ask user to provide one.

yes, I have also seen the various ways already available in the code to encrypt the data, I just have to list them so we decide which one to use

> https://github.com/bromite/bromite/issues/244 is mostly to use the same profile across multiple devices.

that's true

**Comment by csagan5** [comment]
See this: https://android-developers.googleblog.com/2022/03/freeing-up-60-of-storage-for-apps.html

That "archived APK" format is not very far from what is described here for this issue.

**Comment by csagan5** [comment]
What permissions do apps need to be able to access this saved backup file **without** user interaction? As far as I understand they would need `MANAGE_EXTERNAL_STORAGE`, otherwise an user interaction is always necessary.

The choice of encryption must be done thinking about the threat model, there is no "stronger is better"; Chromium contains two usable libraries that can create password-protected ZIP files:
1. https://source.chromium.org/chromium/chromium/src/+/main:third_party/minizip/src/README.md (has AES support)
2. https://source.chromium.org/chromium/chromium/src/+/main:third_party/zlib/contrib/minizip/minizip.md (older version, no AES support, but this one seems not used)
3. https://source.chromium.org/chromium/chromium/src/+/main:third_party/libzip/src/README.md (has AES support)

For our use-case I think AES password protection would not even be needed, but it is there readily available so we should use that.
It would also be possible to use AES directly to encrypt, but using a .zip archive is better for convenience.
